rfd/0125-dynamic-auto-discovery-config.md

```diff
@@ -0,0 +1,406 @@
+
+#### Watcher on `DiscoveryConfig`
+If the `discovery_service` tries to set up a watcher over `DiscoveryConfig` but Teleport Auth/Proxy is not yet aware of this resource, then `discovery_service` will log a warning but continue with the static configuration.
+
+### UX
+
+#### `discovery_service` new service configuration properties
+Users need to add the `selector` property and then a list of label matchers.
+
+Example:
+```yaml
+discovery_service:
+  enabled: "yes"
+  selector:
+    labelA: value1 # matches against a single value
+    labelB: [value2, value3] # matches against multiple values
+    labelC: "*" # matches on all values
+```
+
+#### WebAPI: manage `DiscoveryConfig` resource
+The following endpoints must be create to be used by WebAPI:
+```
+Methods:
+GET .../discoveryconfig
+GET .../discoveryconfig/:id
+POST .../discoveryconfig
+PUT .../discoveryconfig/:id
+DELETE .../discoveryconfig/:id
+```
+
+JSON representation:
+```json
+{
+   "aws":[
+      {
+         "types":[
+            "ec2"
+         ],
+         "regions":[
+            "us-east-1",
+            "us-west-1"
+         ],
+         "tags":{
+            "*":"*"
+         },
+         "install":{
+            "join_params":{
+               "token_name":"aws-discovery-iam-token"
+            },
+            "script_name":"default-installer"
+         },
+         "ssm":{
+            "document_name":"TeleportDiscoveryInstaller"
+         }
+      }
+   ],
+   "azure":[
+      {
+         "types":[
+            "aks"
+         ],
+         "regions":[
+            "eastus",
+            "westus"
+         ],
+         "subscriptions":[
+            "11111111-2222-3333-4444-555555555555"
+         ],
+         "resource_groups":[
+            "group1",
+            "group2"
+         ],
+         "tags":{
+            "*":"*"
+         }
+      }
+   ],
+   "gcp":[
+      {
+         "types":[
+            "gke"
+         ],
+         "locations":[
+            "*"
+         ],
+         "tags":{
+            "*":"*"
+         },
+         "project_ids":[
+            "myproject"
+         ]
+      }
+   ]
+}
+```
+
+#### `tctl`: manage `DiscoveryConfig` resource
+Users must be able to create, list, read, update and delete `DiscoveryConfig`s using `tctl`.
+
+Example of `tctl get discovery_config`:
+```
+$ tctl get discovery_config
+kind: DiscoveryConfig
+version: v1
+metadata:
+  name: production-resources
+  labels:
+    cloud: aws
+    env: prod
+spec:
+    aws:
+     - types: ["ec2"]
+       regions: ["us-east-1","us-west-1"]
+       tags:
+         "*": "*"
+```
+
+#### IaC - Terraform
+A new resource `DiscoveryConfig` must be created to allow its management from the Terraform provider.
+
+#### IaC - Kube Operator
+A new resource `DiscoveryConfig` must be created to allow its management from the Kube Operator.
+
+#### IaC - Helm Charts
+Helm charts must support the new property when setting up a `discovery_service`.
+
+### Security
```

Comment:
Can you address RBAC in this section? What kind of permissions/verbs will be required for users to be able to manage these configs?

Reply:
Yes, I'll add some notes :+1: